### PR TITLE
feat(VRating): port to v3

### LINKED ISSUE #12629: [Bug Report] Ratings props.isHovered isn't working
### Environment
**Vuetify Version:** 2.3.17
**Vue Version:** 2.6.12
**Browsers:** Chrome 86.0.4240.183
**OS:** Mac OS 10.15.7

### Steps to reproduce
Try to access the isHovered prop in the v-ratings item slot.

### Expected Behavior
isHovered should be true with the icon is hovered

### Actual Behavior
the isHovered prop value does not change

### Reproduction Link
<a href="https://codepen.io/marianicolosi/pen/dyXrmEj?editors=1111" target="_blank">https://codepen.io/marianicolosi/pen/dyXrmEj?editors=1111</a>

### PR DESCRIPTION
**Pull request by nekosaur**
## Description
Updated v-rating for v3.

* fixes #12629

## Motivation and Context

v3

## Breaking changes
- Added `tag` prop
- Added `item-labels`, `item-label-position` props and `item-label` slot (https://github.com/vuetifyjs/vuetify/issues/5167)
- Renamed `icon-label` to `item-aria-label`
- Renamed `rating.ariaLabel.icon` to `rating.ariaLabel.item` entry in localization files
- Removed `dense` prop, added `density` prop
- Removed individual size props, added `size` prop
- `item` slot props have changed

## Markup:

<details>

```vue
<template>
  <v-app>
    <v-rating
      v-model="rating"
      background-color="purple-lighten-3"
      color="purple"
      size="large"
      hover
      length="10"
    ></v-rating>
    {{ rating }}
  </v-app>
</template>

<script>
  import { ref } from 'vue'

  export default {
    name: 'Playground',
    setup () {
      const rating = ref()

      return { rating }
    },
  }
</script>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Review by johnleider — CHANGES_REQUESTED** [review]
Everything looks good except for the hover effects (which may be a v-btn thing), they feel delayed/laggy.

rating

File: packages/docs/src/examples/v-rating/slot-item.vue [hunk 1]
```diff
@@ -7,7 +7,7 @@
           large
           @click="props.onClick"
         >
-          {{ props.isFilled ? 'mdi-star-circle' : 'mdi-star-circle-outline' }}
+          mdi-star-circle{{ props.isFilled ? '' : '-outline' }}
```
Comment (nekosaur) [inline packages/docs/src/examples/v-rating/slot-item.vue:10]: not really worth arguing over, but I think that for an example, the old way is actually better/clearer

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 2]
```diff
@@ -204,7 +205,7 @@ export default defineComponent({
               : iconProps.label ? <span>{iconProps.label}</span>
               : <span>&nbsp;</span>
             }
-            {slots.item ? slots.item(iconProps) : (
+            { slots.item ? slots.item(iconProps) : (
```
Comment (nekosaur) [inline packages/vuetify/src/components/VRating/VRating.tsx:207]: what's our rule for space after `{` or not in JSX? You didn't add it for the prop values. And you also didn't add it at the end of this expression.

Comment (johnleider) [reply packages/vuetify/src/components/VRating/VRating.tsx:217]: https://github.com/vuetifyjs/vuetify/blob/next/packages/vuetify/src/components/VBanner/VBanner.tsx#L80 was my inspiration.

File: packages/docs/src/examples/v-rating/slot-item.vue (hunk 1 above)
Comment (johnleider) [reply packages/docs/src/examples/v-rating/slot-item.vue:10]: I'm in agreeance, convert to the previous implementation.

**Comment by johnleider** [comment]
Requesting **Final** comments from @vuetifyjs/core-team

**Review by johnleider — CHANGES_REQUESTED** [review]
Components that have a default color should define it as a sass variable with the prop being used to override that.

`v-btn` default icon styles changed, they now have elevation that should be removed.

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    })))
+
+    function updateFocus () {
+      const index = Math.floor(rating.value - 0.5)
+      const el = refs.value[index]?.$el ?? refs.value[index]
+      el?.focus()
+    }
+
+    function onKeydown (e: KeyboardEvent) {
+      if (![keyCodes.left, keyCodes.right].includes(e.keyCode)) return
```
Comment (KaelWD) [inline packages/vuetify/src/components/VRating/VRating.tsx:190]: Are we still using these? `e.keyCode` is deprecated and everything supports `e.key` (or `e.code`) now.

Comment (albertkaaman1) [reply packages/vuetify/src/components/VRating/VRating.tsx:190]: If it's deprecated then we should not

Comment (nekosaur) [reply packages/vuetify/src/components/VRating/VRating.tsx:190]: If it's deprecated then we should not

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                size={itemProps.size}
+                icon={itemProps.icon}
+                onClick={itemProps.onClick}
+                onMouseenter={itemProps.onMouseenter}
+                onMouseleave={itemProps.onMouseleave}
+                onMousemove={itemProps.onMousemove}
+                ariaLabel={itemProps.ariaLabel}
+                disabled={itemProps.disabled}
+                density={itemProps.density}
+                tabindex={itemProps.tabindex}
```
Comment (KaelWD) [inline packages/vuetify/src/components/VRating/VRating.tsx:243]: These should probably all be spread (`{ ...itemProps }`)

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+      size: props.size,
+      tabindex: props.readonly ? -1 : undefined,
+      text: true,
+      value: rating.value,
+      ...itemState.value[index],
+      ...eventState.value[index],
+    })))
+
+    function updateFocus () {
+      const index = Math.floor(rating.value - 0.5)
```
Comment (KaelWD) [inline packages/vuetify/src/components/VRating/VRating.tsx:184]: Hidden `<input type="radio">` will do this for you, and support screen readers properly. For baseline screen reader support you need at least `role="radio" ariaChecked={ rating.value === index + 1 }` on the buttons.

File: packages/vuetify/src/components/VRating/VRating.sass [hunk 6]
```diff
@@ -1,28 +1,56 @@
...
+    opacity: 1
+
+.v-rating__item--focused:before
+  content: " "
+  position: absolute
+  left: 0px
+  right: 0px
+  top: 0px
+  bottom: 0px
+  border: 1px solid black
```
Comment (johnleider) [inline packages/vuetify/src/components/VRating/VRating.sass:40]: is this intentional?

File: packages/vuetify/src/components/VRating/VRating.sass [hunk 7]
```diff
@@ -1,28 +1,56 @@
...
+  pointer-events: none
+
+.v-rating--hidden
+  position: absolute
+  width: 0px
```
Comment (johnleider) [inline packages/vuetify/src/components/VRating/VRating.sass:45]: I'm assuming here the **px** is intentional

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+// Types
+import type { Prop } from 'vue'
+
+export default defineComponent({
+  name: 'VRating',
+
+  props: makeProps({
+    name: {
+      type: String,
+      required: true,
```
Comment (johnleider) [inline packages/vuetify/src/components/VRating/VRating.tsx:30]: we should auto generate one for the user, requiring a name is burdensome.

**Review by johnleider — CHANGES_REQUESTED** [review]

File: packages/vuetify/src/composables/proxiedModel.ts [hunk 9]
```diff
@@ -28,7 +28,7 @@ export function useProxiedModel<
     )
   })
 
-  const internal = ref(transformIn(propIsDefined.value ? props[prop] : defaultValue)) as Ref<Inner>
+  const internal = ref(transformIn(props[prop])) as Ref<Inner>
```
Comment (KaelWD) [inline packages/vuetify/src/composables/proxiedModel.ts:31]: What was this change for? It breaks the snapshots.

File: packages/vuetify/src/composables/refs.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+// Imports
+import { onBeforeUpdate, ref } from 'vue'
+
+export function useRefs <T extends {}> () {
```
Comment (KaelWD) [inline packages/vuetify/src/composables/refs.ts:4]: Unused?

File: packages/vuetify/src/components/VRating/_variables.scss [hunk 11]
```diff
@@ -1,5 +1,15 @@
...
-$rating-dense-padding: 0.1rem !default;
+
+$rating-sizes: () !default;
+$rating-sizes: map-deep-merge(
+  (
+    'x-small': 16px,
+    'small': 20px,
+    'default': 24px,
+    'large': 28px,
+    'x-large': 32px,
```
Comment (KaelWD) [inline packages/vuetify/src/components/VRating/_variables.scss:12]: VBtn just defines a base size then the others are calculated from `$size-scales` &times; some multiplier

File: packages/vuetify/src/composables/proxiedModel.ts (hunk 9 above)
Comment (nekosaur) [reply packages/vuetify/src/composables/proxiedModel.ts:31]: might be experiment I forgot to revert

File: packages/vuetify/src/components/VRating/_variables.scss (hunk 11 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VRating/_variables.scss:12]: will remove, as current implementation just renders v-btns

**Review by johnleider — APPROVED** [review]

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+    disabled: Boolean,
+    emptyIcon: {
+      type: String,
+      default: '$ratingEmpty',
+    },
+    fullIcon: {
+      type: String,
+      default: '$ratingFull',
+    },
+    halfIncrements: Boolean,
```
Comment (nekosaur) [inline packages/vuetify/src/components/VRating/VRating.tsx:44]: have you looked at focused state when using half-increments?

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+              : slots.item ? slots.item({
+                ...itemState.value,
+                props: btnProps,
+                value,
+                index,
+              })
+              : (
+                <VBtn
+                  color={ itemState.value[index].color }
+                  icon={ itemState.value[index].icon }
```
Comment (nekosaur) [inline packages/vuetify/src/components/VRating/VRating.tsx:177]: these should be part of btnProps

File: packages/vuetify/src/components/VRating/VRating.tsx [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+    function onMouseup () {
+      isClicking = false
+    }
+
+    const name = computed(() => props.name ?? `v-rating-${getUid()}`)
+
+    function VRatingItem ({ value, index, showStar = true }: { value: number, index: number, showStar?: boolean }) {
+      const { onMouseenter, onMouseleave, onFocus, onBlur, onClick } = eventState.value[index + 1]
+      const id = `${name.value}-${String(value).replace('.', '-')}`
+      const btnProps = {
```
Comment (nekosaur) [inline packages/vuetify/src/components/VRating/VRating.tsx:144]: you're not passing disabled prop

File: packages/vuetify/src/composables/refs.ts (hunk 10 above)
Comment (nekosaur) [reply packages/vuetify/src/composables/refs.ts:4]: yes

Comment (nekosaur) [reply packages/vuetify/src/composables/refs.ts:4]: yes

Comment (johnleider) [reply packages/vuetify/src/composables/refs.ts:4]: Leave in or take out?

**Pull request merged**